http/README.md

```diff
@@ -0,0 +1,159 @@
+
+This document defines how libp2p nodes can offer a HTTP endpoint next to (or instead of) their full libp2p node. Services can be offered both via traditional libp2p protocols and via HTTP, allowing a wide variety of nodes to access these services. Crucially, this for the first time, allows browsers to access libp2p services without spinning up a Web{Socket, Transport, RTC} connection first. It also allows interacting with libp2p services from environments where plain HTTP is the only option, e.g. curl from the command line, and certain cloud edge workers and lambdas.
+
+At the same time, nodes that are already connected via a libp2p connection, will be able to (re)use this connection to issue the same kind of requests, without dialing a dedicated HTTP connection.
+
+Any protocol that follows request-response semantics can easily be mapped onto HTTP (mapping protocols that don’t follow a request-response flow can be more challenging). Protocols are encouraged to follow best practices for building REST APIs. Once a mapping has been defined, a single implementation can be used to serve both traditional libp2p as well as libp2p-HTTP clients.
+
+## Addressing
+
+Nodes may advertise HTTP multiaddresses to signal support for libp2p over HTTP. An address might look like this: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/p2p/<peer id>` (for HTTP/1.1 and HTTP/2), or  `/ip4/1.2.3.4/udp/443/quic/sni/example.com/http/p2p/<peer id>` (for HTTP/3).
+
+Nodes MUST use HTTPS (i.e. they MUST NOT use unencrypted HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3, but the protocols also work over HTTP/1.1.
+
+Note that the peer ID in this address is 1. optional and 2. advisory and not (necessarily) verified during the HTTP handshake (depending on the HTTP client). If and when desired, clients can cryptographically verify the peer ID once the HTTP connection has been established, see [Authentication] for details on peer authentication.
+
+Nodes can also link to a specific resource directly, similar to how a URL includes a path. This will require us to resolve [https://github.com/multiformats/multiaddr/issues/63](https://github.com/multiformats/multiaddr/issues/63) first. For example, the URL of a specific CID might be: `/ip4/1.2.3.4/tcp/443/tls/sni/example.com/http/<my data transfer protocol>/{/path/to/<cid>}`.
+
+## Namespace
+
+libp2p does not squat the global namespace. By convention, all libp2p services are located at a well-known URL: `http://example.com/.well-known/libp2p/<service name>/<path (optional)>`.
+
+Putting the service name into the URL allows for future extensibility. It is easy to define new protocols, and the replace existing protocols by newer versions.
+
+Applications MAY expose services under different URIs. For example, an application might decide to generate nicer-looking (and probably more SEO-friendly) URLs, and map paths under `[https://example.com/dht/](https://example.com/dht/)` to `https://example.com/.well-known/libp2p/kad-dht-v1/`. 
```

Comment:
1. We need to have HTTP endpoints for libp2p-internal protocols. This includes Identify (or a split-up version of that protocol, as @thomaseizinger suggested), Kademlia, Gossipsub (?), etc. At the very least, we need to decide where these protocols are available.
2. Who gets to decide where protocols are mapped? A typical stack will have multiple layers of protocols. If they all start polluting the global namespace, it won't be possible to run a normal webserver any more, since the URL endpoints will be taken. In my view, _only_ the top layer should make decisions about the global namespace.

Reply:
We could leverage links here:

1. Protocols define paths. Let's say, kademlia has the `/kademlia/values` path and accepts `PUT` requests there.
2. Applications can mount the paths of a libp2p library anywhere they want.
3. We define an index document that MUST sit in a well-known location, i.e. `example.com/.well-known/libp2p`.
4. Said index document lays out, where to find the libp2p paths (and maybe other metadata):

```json
{
	"version": "0.25.0",
	"mounted_at": "/libp2p"
}
```
5. Implementations could then discover the mount point and automatically prefix the `/libp2p` segment. The kademlia module would then make requests to `/libp2p/kademlia/values`.

This way, the global namespace is not squatted and protocols at the same time have freedom over their URL design.

Reply:
An alternative to `./well-known/libp2p` would be to tell applications to include a `Link` header in their index response. Implementations would then:

1. Make a `HEAD` request to `/`
2. Search for a `Link` header with `rel=libp2p`, i.e. `Link: /libp2p; rel="libp2p"`

Even if said application hosts many many different websites and endpoints, it shouldn't be too much to ask to include a `Link` header in a GET/HEAD request to `/`.

Reply:
If I remember correctly, then per web linking spec (RFC 8288), you cannot just invent link relations and they should be registered with IANA instead so we can't just use `libp2p`. But, the spec also defines what is called an extended link relation. Those are links (inception anyone?) that MAY resolve to a document explaining the link relation. Something like:

`Link: </libp2p-sub-path>; rel="https://docs.libp2p.io/link-relations/mount-point"`

Going to https://docs.libp2p.io/link-relations/mount-point could then show a piece of documentation that explains what implementations should do here.

Reply:
All of this is at the cost of 1 round-trip when you first talk to a given libp2p node, assuming we cache the response.